### PR TITLE
Plugin lifecycle startup/shutdown not called.

### PR DESCRIPTION
**Pull request by garypen**
Plugin lifecycle startup/shutdown not called.
    
    First attempt at implementing a minimal version of plugin lifecycle
    support.
    
    Key points:
     - Add a trait to Plugin fn name() -> &'static str which makes it easy
       to identify plugins (for logging) at runtime. May be replaced in long term
       by a different mechanism, but good enough for now.
     - Add an interface to the PluggableRouterServiceBuilder so that we can
       consume the builder and retrieve the plugins which have been
       processed so far. Required to help cleanup started plugins.
     - Also, modify the build() method to return both the service and the
       set of plugins used to build the service. This is then maintained
       within our configuration builder and used to shut down old plugins on
       configuration changes or shutdown.
     - In the builder, no longer reverse the order of plugins, but reverse the
       iterators so we can return the plugins in the original order at the end
       of the function.
     - I've added a couple of errors to the ConfigurationErrors to represent
       plugins failing to start or stop. Stop isn't currently used, so may
       delete before final commit.
     - I've added a pile of plugins for testing purposes.
     - I've added Error and Display and Debug implementations/derives to the
       Hello plugin and implemented startup/shutdown methods.
     - I register the hello plugin under different names to ensure that is
       working properly with respect to plugin ordering in ad-hoc tests.
     - Because I now maintain the set of plugins used to build a service in
       the YamlRouterServiceFactory, this makes various method &mut rather
       than &. This propagates into the state machine. Not ideal, but better
       than handling internal mutability I think. Suggestions about better
       ways to achieve this are welcome.
     - I've implemented Drop on the YamlRouterServiceFactory. This cleans up
       plugins on shutdown of the router.
     - When plugins register, their startup method is now invoked.
     - If errors in config/startup are detected, any started plugins are
       shutdown
     - If a re-config is clean, then any previous plugins (i.e. from last
       configuration event, guaranteed empty at startup) are shutdown.
       Because we drain the list here, if we then later shutdown the router,
       the shutdown isn't performed again.
     - I've changed the "group" to "name" separator from "_" to ".".
    
    This work is all focussed on DynPlugin lifecycles. I have a slight
    concern that I may be missing something relating to non-dynamic plugins,
    but I'll need to speak to someone about that once this has been
    reviewed.
    
    This work doesn't attempt to address the "what happens if a plugin
    hangs" aspects of lifecycle management.

(I've also improved spaceport logging in here. Not part of this work, but I didn't want to lose the improvements.)

fixes: #409

### CODE REVIEW

File: apollo-router-core/src/plugin.rs [hunk 1]
```diff
@@ -103,7 +104,7 @@ pub trait Plugin: Send + Sync + 'static + Sized {
 }
 
 #[async_trait]
-pub trait DynPlugin: Send + Sync + 'static {
+pub trait DynPlugin: Display + Send + Sync + 'static {
```
Comment (bnjjj) [inline apollo-router-core/src/plugin.rs:107]: I think this trait bound is a little bit annoying for people who just want to implement a plugin. And also requiring Display just to display the name is a little bit weird to me. Maybe we could add a method in the plugin trait like `fn name() -> &'static str` of equivalent. It's less restrictive and pretty straightforward. And with this function in the trait we could get rid of passing the name in the `register_plugin` macro. What do you think ?

File: apollo-router/src/plugins/hello.rs [hunk 2]
```diff
@@ -1,25 +1,60 @@
...
+#[derive(Debug)]
+struct Hello {
+    name: String,
+}
+
+impl fmt::Display for Hello {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Hello")
+    }
+}
```
Comment (bnjjj) [inline apollo-router/src/plugins/hello.rs:28]: We can delete this part now I guess

Comment (garypen) [reply apollo-router/src/plugins/hello.rs]: done: 022587d95a1a1060e126045967cef68efae90c48

File: apollo-router/src/router_factory.rs [hunk 3]
```diff
@@ -154,9 +193,181 @@ mod test {
...
+    // Always starts and stops plugin
+
+    #[derive(Debug)]
+    struct AlwaysStartsAndStopsPlugin {}
+
+    impl fmt::Display for AlwaysStartsAndStopsPlugin {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "AlwaysStartsAndStopsPlugin")
+        }
+    }
```
[the remark this answers is not in the document]
Comment (garypen) [reply apollo-router/src/router_factory.rs]: done: 022587d95a1a1060e126045967cef68efae90c48

File: apollo-router/src/router_factory.rs [hunk 4]
```diff
@@ -154,9 +193,181 @@ mod test {
...
+    // Always fails to start plugin
+
+    #[derive(Debug)]
+    struct AlwaysFailsToStartPlugin {}
+
+    impl fmt::Display for AlwaysFailsToStartPlugin {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "AlwaysFailsToStartPlugin")
+        }
+    }
```
[the remark this answers is not in the document]
Comment (garypen) [reply apollo-router/src/router_factory.rs]: done: 022587d95a1a1060e126045967cef68efae90c48

File: apollo-router/src/router_factory.rs [hunk 5]
```diff
@@ -154,9 +193,181 @@ mod test {
...
+    // Always fails to stop plugin
+
+    #[derive(Debug)]
+    struct AlwaysFailsToStopPlugin {}
+
+    impl fmt::Display for AlwaysFailsToStopPlugin {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "AlwaysFailsToStopPlugin")
+        }
+    }
```
[the remark this answers is not in the document]
Comment (garypen) [reply apollo-router/src/router_factory.rs]: done: 022587d95a1a1060e126045967cef68efae90c48

**Review by bnjjj — APPROVED** [review]

**Review by BrynCooke — APPROVED** [review]
Approved pending a comment about why we exit if plugins fail to shut down.

**Comment by Geal** [comment]
exiting looks dangerous yes. That should raise a big alert, but I'd prefer the router runs in a degraded state (with a plugin leaking memory or CPU) instead of stopping entirely. Stopping should be fine in debug mode though

**Comment by garypen** [comment]
> exiting looks dangerous yes. That should raise a big alert, but I'd prefer the router runs in a degraded state (with a plugin leaking memory or CPU) instead of stopping entirely. Stopping should be fine in debug mode though

We don't know why the plugin failed. We only know it failed. To be safe, which I think is the important thing here, we must assume the worst and exit the process with useful debugging details for post-mortem analysis.

If we decide that we want to have error classes for plugins, then we could introduce such a mechanism and use that to determine how to interpret plugin errors. I feel like that would be overkill and we can always introduce that later.

Also: The general principle I'm following here is that it's better to be strict initially and relax later if justified.

**Comment by Geal** [comment]
Alright then 🙂

**Review by Geal — APPROVED** [review]

File: apollo-router-core/src/plugin.rs [hunk 6]
```diff
@@ -187,7 +201,7 @@ macro_rules! register_plugin {
                 $name.to_string()
             }
             else {
-                format!("{}_{}", $group, $name)
+                format!("{}.{}", $group, $name)
```
Comment (o0Ignition0o) [inline apollo-router-core/src/plugin.rs:204]: 🎉 YES!

**Review by o0Ignition0o — APPROVED** [review]

**Pull request merged**